exec/bigmachine.go

```diff
@@ -95,12 +96,16 @@ type bigmachineExecutor struct {
 	// slices on bigmachine workers. Note that this requires that we
 	// hold on to the invocations, which is somewhat unfortunate, but
 	// I don't see a clean way around it.
-	invocations    map[uint64]bigslice.Invocation
+	invocations    map[uint64]execInvocation
 	invocationDeps map[uint64]map[uint64]bool
 
-	// compileEnvs maintains the compilation environment used to compile tasks
-	// (for a particular invocation index).
-	compileEnvs map[uint64]CompileEnv
+	// invocationBuffers holds the gob-encoded representations of the
+	// corresponding invocations. Because Func arguments, held in invocations,
+	// may be large, we memoize the encoded versions so that we don't pay the
+	// cost of gob-encoding the invocations for each worker (CPU to encode;
+	// memory for ephemeral buffers in gob). Instead, we do it once and reuse
+	// the result for each worker.
+	invocationBuffers map[uint64]bytes.Buffer
```

Comment:
Optional: Since these are write-once, consider using `[]byte` instead, since `bytes.Buffer` suggests mutation. In fact, we could even use `string` to indicate immutability, but I'm not sure what most Go readers' expectations are about "real" strings. (`bytes.Reader` or `strings.Reader` will allow streaming the argument.)

Optional: Maybe `Discard` should delete this entry.

Reply:
I agree that using `[]byte` is better. I will do that.

I also agree that `Discard` could/should delete the entry, but I'll save that for another pass to get this merged more quickly.